app/master/build_request_handler.py

```diff
@@ -101,6 +101,11 @@ def _prepare_build_async(self, build, project_lock):
                                    log_msg='Build preparation loop is handling request for build {build_id}.')
             try:
                 build.prepare(self._subjob_calculator)
+                # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                if len(build.all_subjobs()) == 0:
+                    self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                    build.mark_started()
+                    build.perform_async_postbuild_tasks()
```

Comment:
I was contemplating creating a "mark_completed()" function in the Build class, but it would end up entirely just being a wrapp for the perform_async_postbuild_tasks() invocation, so I decided to make this method public instead.